placement/algo/sharded_helper.go

```diff
@@ -738,3 +738,80 @@ func newShards(shardIDs []uint32) []shard.Shard {
 	}
 	return r
 }
+
+func removeInstanceFromList(list []placement.Instance, instanceID string) []placement.Instance {
+	for i, instance := range list {
+		if instance.ID() == instanceID {
+			last := len(list) - 1
+			list[i] = list[last]
+			return list[:last]
+		}
+	}
+	return list
+}
+
+func markShardAvailable(p placement.Placement, instanceID string, shardID uint32, nowNanos int64, shouldCheckCutoverTime bool) (placement.Placement, error) {
+	instance, exist := p.Instance(instanceID)
+	if !exist {
+		return nil, fmt.Errorf("instance %s does not exist in placement", instanceID)
+	}
+
+	shards := instance.Shards()
+	s, exist := shards.Shard(shardID)
+	if !exist {
+		return nil, fmt.Errorf("shard %d does not exist in instance %s", shardID, instanceID)
+	}
+
+	if s.State() != shard.Initializing {
+		return nil, fmt.Errorf("could not mark shard %d as available, it's not in Initializing state", s.ID())
+	}
+
+	if shouldCheckCutoverTime && s.CutoverNanos() > nowNanos {
+		return nil, fmt.Errorf("could not mark shard %d as available, it's cutover time %d is scheduled after now %d", s.ID(), s.CutoverNanos(), nowNanos)
+	}
+
+	sourceID := s.SourceID()
+	shards.Add(shard.NewShard(shardID).SetState(shard.Available))
+
+	// There could be no source for cases like initial placement.
+	if sourceID == "" {
+		return p, nil
+	}
+
+	sourceInstance, exist := p.Instance(sourceID)
+	if !exist {
+		return nil, fmt.Errorf("source instance %s for shard %d does not exist in placement", sourceID, shardID)
+	}
+
+	sourceShards := sourceInstance.Shards()
+	sourceShard, exist := sourceShards.Shard(shardID)
+	if !exist {
+		return nil, fmt.Errorf("shard %d does not exist in source instance %s", shardID, sourceID)
+	}
+
+	if sourceShard.State() != shard.Leaving {
+		return nil, fmt.Errorf("shard %d is not leaving instance %s", shardID, sourceID)
+	}
+
+	sourceShards.Remove(shardID)
+	if sourceShards.NumShards() == 0 {
+		return p.SetInstances(removeInstanceFromList(p.Instances(), sourceInstance.ID())), nil
+	}
+	return p, nil
+}
+
+func markAllShardsAvailable(p placement.Placement, shouldCheckCutoverTime bool, nowNanos int64) (placement.Placement, error) {
```

Comment:
We do, this is mostly used in BuildInititialPlacement(), all the shard in the initial placement will all have cutover time in the future rather than 0.

Reply:
Is that necessary? I'd imagine the initial placement is for setting up the agg tier for some initial testing, so might be better to get feedback sooner than waiting till the cutover time. Not feeling strongly about it either way though.

Reply:
This means the generated initial placement will have all shards marked as available, the markShardAsAvailable was done in placement service and I moved that to the algo in this diff

Reply:
Oh I see cool sounds good.